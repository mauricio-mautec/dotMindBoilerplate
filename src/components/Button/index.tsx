import { forwardRef, AnchorHTMLAttributes, ButtonHTMLAttributes } from 'react'
import * as S from './styles'
/*
 https://www.saltycrane.com/cheat-sheets/typescript/react/latest/
    interface AreaHTMLAttributes<T> extends HTMLAttributes<T> {
        alt?: string;
        coords?: string;
        download?: any;
        href?: string;
        hrefLang?: string;
        media?: string;
        rel?: string;
        shape?: string;
        target?: string;
    }
    interface ButtonHTMLAttributes<T> extends HTMLAttributes<T> {
        autoFocus?: boolean;
        disabled?: boolean;
        form?: string;
        formAction?: string;
        formEncType?: string;
        formMethod?: string;
        formNoValidate?: boolean;
        formTarget?: string;
        name?: string;
        type?: 'submit' | 'reset' | 'button';
        value?: string | string[] | number;
    }
 */
type ButtonTypes =
  | AnchorHTMLAttributes<HTMLAnchorElement>
  | ButtonHTMLAttributes<HTMLButtonElement>

export type ButtonProps = {
  size?: 'small' | 'medium' | 'large'
  fullWidth?: boolean
  minimal?: boolean
  icon?: JSX.Element
  as?: React.ElementType
} & ButtonTypes

const Button: React.ForwardRefRenderFunction<S.WrapperProps, ButtonProps> = (
  {
    children,
    icon,
    size = 'medium',
    fullWidth = false,
    minimal = false,
    ...props
  },
  ref
) => (
  <S.Wrapper
    size={size}
    fullWidth={fullWidth}
    hasIcon={!!icon}
    minimal={minimal}
    ref={ref}
    {...props}
  >
    {!!icon && icon}
    {!!children && <span>{children}</span>}
  </S.Wrapper>
)

export default forwardRef(Button)
