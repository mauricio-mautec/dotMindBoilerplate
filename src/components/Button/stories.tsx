import { ComponentStory, ComponentMeta } from '@storybook/react'
import { AddShoppingCart } from '@styled-icons/material-outlined/AddShoppingCart'

import Button from '.'

export default {
  title: 'Button',
  component: Button
} as ComponentMeta<typeof Button>

const Template: ComponentStory<typeof Button> = (args) => <Button {...args} />

export const Basic = Template.bind({})
export const withIcon = Template.bind({})
export const asLink = Template.bind({})
export const asDisabledButton = Template.bind({})

Basic.args = {
  children: 'Buy Now',
  size: 'medium'
}
Basic.argTypes = {
  icon: {
    table: {
      disable: true
    }
  },
  as: {
    control: 'select',
    options: ['a', 'button']
  }
}

withIcon.args = {
  children: 'Buy Now Icon',
  size: 'small',
  icon: <AddShoppingCart />
}
withIcon.argTypes = {
  icon: {
    table: {
      disable: true
    }
  },
  as: {
    control: 'select',
    options: ['a', 'button']
  }
}

asLink.args = {
  size: 'large',
  children: 'Buy now',
  as: 'a',
  href: '/link'
}
asLink.argTypes = {
  icon: {
    table: {
      disable: true
    }
  },
  as: {
    table: { disable: true }
  }
}
asDisabledButton.args = {
  size: 'large',
  children: 'Buy now',
  as: 'button',
  href: '/link',
  disabled: true
}
asDisabledButton.argTypes = {
  icon: {
    table: {
      disable: true
    }
  },
  as: {
    table: { disable: true }
  }
}
