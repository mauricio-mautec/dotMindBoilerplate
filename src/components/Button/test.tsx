import { screen } from '@testing-library/react'
import { renderWithTheme } from 'utils/tests/helpers'
import { AddShoppingCart } from '@styled-icons/material-outlined/AddShoppingCart'

import Button from '.'

describe('<Button />', () => {
  it('should render the medium size by default', () => {
    const { container } = renderWithTheme(<Button>Buy Now</Button>)
    expect(screen.getByRole('button', { name: /buy now/i })).toHaveStyle({
      height: '4rem',
      padding: '0.8rem 3.2rem',
      'font-size': '1.4rem'
    })
    expect(container.firstChild).toMatchSnapshot('default')
  })
  it('should render the small size', () => {
    const { container } = renderWithTheme(<Button size="small">Buy Now</Button>)
    expect(screen.getByRole('button', { name: /buy now/i })).toHaveStyle({
      height: '3rem',
      padding: '0.8rem',
      'font-size': '1.2rem'
    })
    expect(container.firstChild).toMatchSnapshot('small')
  })
  it('should render the large size', () => {
    const { container } = renderWithTheme(<Button size="large">Buy Now</Button>)
    expect(screen.getByRole('button', { name: /buy now/i })).toHaveStyle({
      height: '5rem',
      padding: '0.8rem 4.8rem',
      'font-size': '1.6rem'
    })
    expect(container.firstChild).toMatchSnapshot('large')
  })
  it('should render a fullWidth version', () => {
    const { container } = renderWithTheme(<Button fullWidth>Buy Now</Button>)
    expect(screen.getByRole('button', { name: /buy now/i })).toHaveStyle({
      width: '100%'
    })
    expect(container.firstChild).toMatchSnapshot('fullwidth')
  })
  it('should render a icon version', () => {
    const { container } = renderWithTheme(
      <Button icon={<AddShoppingCart data-testid="icon" />}>Buy Now</Button>
    )
    expect(screen.getByText(/buy now/i)).toBeInTheDocument()
    expect(screen.getByTestId('icon')).toBeInTheDocument()
    expect(container.firstChild).toMatchSnapshot('icon')
  })
})
