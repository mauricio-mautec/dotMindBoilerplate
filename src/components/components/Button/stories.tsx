import { ComponentStory, ComponentMeta } from '@storybook/react'
import { AddShoppingCart } from '@styled-icons/material-outlined/AddShoppingCart'

import Button from '.'

export default {
  title: 'Button',
  component: Button
} as ComponentMeta<typeof Button>

const Template: ComponentStory<typeof Button> = (args) => <Button {...args} />

export const Basic = Template.bind({})
export const withIcon = Template.bind({})

Basic.args = {
  children: 'Buy Now',
  size: 'medium'
}

withIcon.args = {
  children: 'Buy Now Icon',
  size: 'small',
  icon: <AddShoppingCart />
}
