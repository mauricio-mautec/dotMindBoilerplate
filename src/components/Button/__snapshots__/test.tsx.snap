// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button /> should render a fullWidth version: fullwidth 1`] = `
.c0 {
  background: linear-gradient(180deg,#ff5f5f 0%,#f062c0 50%);
  color: #FAFAFA;
  border: 0;
  cursor: pointer;
  border-radius: 0.4rem;
  height: 4rem;
  font-size: 1.4rem;
  padding: 0.8rem 3.2rem;
  width: 100%;
}

<button
  class="c0"
>
  <span>
    Buy Now
  </span>
</button>
`;

exports[`<Button /> should render a icon version: icon 1`] = `
.c1 {
  display: inline-block;
  vertical-align: middle;
  overflow: hidden;
}

.c0 {
  background: linear-gradient(180deg,#ff5f5f 0%,#f062c0 50%);
  color: #FAFAFA;
  border: 0;
  cursor: pointer;
  border-radius: 0.4rem;
  height: 4rem;
  font-size: 1.4rem;
  padding: 0.8rem 3.2rem;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

.c0 svg {
  width: 1.5rem;
}

.c0 svg + span {
  margin-left: 0.8rem;
}

<button
  class="c0"
>
  <svg
    aria-hidden="true"
    class="c1"
    data-testid="icon"
    fill="currentColor"
    focusable="false"
    viewBox="0 0 24 24"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M0 0h24v24H0V0z"
      fill="none"
    />
    <path
      d="M11 9h2V6h3V4h-3V1h-2v3H8v2h3v3zm-4 9c-1.1 0-1.99.9-1.99 2S5.9 22 7 22s2-.9 2-2-.9-2-2-2zm10 0c-1.1 0-1.99.9-1.99 2s.89 2 1.99 2 2-.9 2-2-.9-2-2-2zm-8.9-5h7.45c.75 0 1.41-.41 1.75-1.03l3.86-7.01L19.42 4l-3.87 7H8.53L4.27 2H1v2h2l3.6 7.59-1.35 2.44C4.52 15.37 5.48 17 7 17h12v-2H7l1.1-2z"
    />
  </svg>
  <span>
    Buy Now
  </span>
</button>
`;

exports[`<Button /> should render the large size: large 1`] = `
.c0 {
  background: linear-gradient(180deg,#ff5f5f 0%,#f062c0 50%);
  color: #FAFAFA;
  border: 0;
  cursor: pointer;
  border-radius: 0.4rem;
  height: 5rem;
  font-size: 1.6rem;
  padding: 0.8rem 4.8rem;
}

<button
  class="c0"
>
  <span>
    Buy Now
  </span>
</button>
`;

exports[`<Button /> should render the medium size by default: default 1`] = `
.c0 {
  background: linear-gradient(180deg,#ff5f5f 0%,#f062c0 50%);
  color: #FAFAFA;
  border: 0;
  cursor: pointer;
  border-radius: 0.4rem;
  height: 4rem;
  font-size: 1.4rem;
  padding: 0.8rem 3.2rem;
}

<button
  class="c0"
>
  <span>
    Buy Now
  </span>
</button>
`;

exports[`<Button /> should render the small size: small 1`] = `
.c0 {
  background: linear-gradient(180deg,#ff5f5f 0%,#f062c0 50%);
  color: #FAFAFA;
  border: 0;
  cursor: pointer;
  border-radius: 0.4rem;
  height: 3rem;
  font-size: 1.2rem;
  padding: 0.8rem;
}

<button
  class="c0"
>
  <span>
    Buy Now
  </span>
</button>
`;
